cfi_backward.adoc

```diff
@@ -128,103 +152,172 @@ instructions on such machines is not supported.
 A push operation is defined as decrement of the `ssp` by `XLEN` followed by a
 write of the link register at the new top of the shadow stack. A pop operation
 is defined as a `XLEN` wide read from the current top of the shadow stack
-followed by an increment of the `ssp`.
+followed by an increment of the `ssp` by `XLEN`.
+
+Usually programs with a shadow stack push the return address onto the regular
+stack as well as the shadow stack in the function prologue. Such programs when
+returning from the function pop the link register from the data stack and pop a
+shadow copy of the link register from the shadow stack. The two values are then
+compared. If the values do not match it is indicative of a corruption of the
+return address variable and the program causes an illegal instruction exception.
+
+To push a link register on the shadow stack, the CFI extension provides a
+`sspush` instruction. 
 
-To push a link register on the shadow stack, the CFI extension provides
-`sspush` instruction. To pop a link register from the shadow stack, the CFI
-extension provides `sspop` instruction.
+A `sspopchk` instruction is provided to pop the shadow return address value
+from the shadow stack and check that the value matches the contents of the
+link register. This instruction is expected to be used by programs operating in
+shadow stack mode.
+
+The CFI extension additionally provides a `sspop` instruction to pop a return
```

Comment:
It's a nit but this paragraph reads a little funny now that it follows `sspopchk` since the functionality is so similar.

Reply:
Its quite different - the `sspop` only pops the return address and is to be used in control stack mode of operation. The `sspopchk` pops return address and checks the return address and is to be used in shadow stack mode of operation. The note below explains the shadow stack and control stack mode